util/compression_test.cc

```diff
@@ -0,0 +1,204 @@
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Testing the features of auto skip compression manager
+//
+// ***********************************************************************
+// EXPERIMENTAL - subject to change while under development
+// ***********************************************************************
+
+#include <cstdlib>
+#include <memory>
+
+#include "db/db_test_util.h"
+#include "port/stack_trace.h"
+#include "rocksdb/flush_block_policy.h"
+#include "table/block_based/block_builder.h"
+#include "test_util/testutil.h"
+#include "util/random.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class AutoSkipTestFlushBlockPolicy : public FlushBlockPolicy {
+ public:
+  explicit AutoSkipTestFlushBlockPolicy(const int window,
+                                        const BlockBuilder& data_block_builder,
+                                        std::shared_ptr<Statistics> statistics)
+      : window_(window),
+        num_keys_(0),
+        data_block_builder_(data_block_builder),
+        statistics_(statistics) {}
+
+  bool Update(const Slice& /*key*/, const Slice& /*value*/) override {
+    auto multiple_of_10 = num_keys_ / window_;
+    if (data_block_builder_.empty()) {
+      // First key in this block
+      return false;
+    }
+    // Check every window
+    if (num_keys_ % window_ == 0) {
+      auto set_exploration = [&](void* arg) {
+        bool* exploration = static_cast<bool*>(arg);
+        *exploration = true;
+      };
+      auto unset_exploration = [&](void* arg) {
+        bool* exploration = static_cast<bool*>(arg);
+        *exploration = false;
+      };
+      SyncPoint::GetInstance()->DisableProcessing();
+      SyncPoint::GetInstance()->ClearAllCallBacks();
+      if (multiple_of_10 % 2 == 0) {
+        SyncPoint::GetInstance()->SetCallBack(
+            "AutoSkipCompressorWrapper::CompressBlock::exploitOrExplore",
+            set_exploration);
+      } else {
+        SyncPoint::GetInstance()->SetCallBack(
+            "AutoSkipCompressorWrapper::CompressBlock::exploitOrExplore",
+            unset_exploration);
+      }
+      SyncPoint::GetInstance()->EnableProcessing();
+
+      auto compressed_count = PopStat(NUMBER_BLOCK_COMPRESSED);
+      auto bypassed_count = PopStat(NUMBER_BLOCK_COMPRESSION_BYPASSED);
+      auto rejected_count = PopStat(NUMBER_BLOCK_COMPRESSION_REJECTED);
+      auto total = compressed_count + rejected_count + bypassed_count;
+      int rejection_percentage, bypassed_percentage, compressed_percentage;
+      if (total != 0) {
+        rejection_percentage = static_cast<int>(rejected_count * 100 / total);
+        bypassed_percentage = static_cast<int>(bypassed_count * 100 / total);
+        compressed_percentage =
+            static_cast<int>(compressed_count * 100 / total);
+      }
+      // use mulitple of 10 to get correct assertion
```

Comment:
Instead of this comment, may be comment on each case what are you trying to verify.